Fix #2447: CancelledError inherits from BaseException on > py 3.8

e-carlin opened the pull request:
Change the code to:
- Catch CancelledError explicitly when needed in cases where we were relying on it to be a subclass of Exception previously.
- Remove any code that does nothing but raise on CancelledError. This pattern was used to not smother CancelledError in cases where we had an `except Exception`. Those will no longer smother the error.

In addition, in sirepo.job_supervisro.run re-raise the CancelledError and remove check for return of False from _send_op. This has no actual behavior change. We don't await _run so it doesn't matter if the CancelledError makes it out. But, it is more "correct" to not smother CancelledError and have it propagate all the way out so a cancel happens fully.

e-carlin commented:
This depends on https://github.com/radiasoft/sirepo/pull/5223 which is waiting on the fedora 36 migration. So, tests won't currently pass.

robnagler commented on sirepo/job_driver/__init__.py:
```diff
@@ -273,7 +274,7 @@ async def _agent_start(self, op):
                 await self.kill()
                 await self._do_agent_start(op)
-            except Exception as e:
+            except (Exception, *sirepo.util.ASYNC_CANCELED_ERROR) as e:
```
I read about `except*`. This nested/multi-exception thing seems very complicated. Do we need it here? It seems like these are two unrelated exceptions.

e-carlin replied:
2f7887c01

robnagler commented on sirepo/job_driver/docker.py:
```diff
@@ -124,10 +124,6 @@ async def kill(self):
             await self._cmd(
                 ("stop", "--time={}".format(job_driver.KILL_TIMEOUT_SECS), self._cname),
             )
-        except sirepo.util.ASYNC_CANCELED_ERROR:
```
This is an implicit coupling on the meaning of ASYNC_CANCELED_ERROR. concurrent.futures.CancelledError is a subclass of Exception so this changes the semantics. Is concurrent.futures.CancelledError not called in 3.10? If it is, then we need to handle that case specially to re-raise like it was. 

I think ASYNC_CANCELED_ERROR belongs in const, but it could be that it should be removed and carefully coded in all cases.

e-carlin replied:
2f7887c01

robnagler commented on sirepo/tornado.py:
```diff
@@ -51,6 +51,7 @@ async def get(self):
             if x:
                 try:
                     r = x.result()
+                # TODO(e-carlin): I think this shouldn't catch CancelledError too. In py 3.8 it won't which I think is the right behavior
```
I think it needs to catch and reraise.

e-carlin replied:
2f7887c01

robnagler commented on sirepo/job_supervisor.py:
```diff
@@ -912,8 +912,7 @@ async def _send_op(op, compute_job_serial, prev_db):
 
         op.task = asyncio.current_task()
         op.pkdel("run_callback")
-        if not await _send_op(op, compute_job_serial, prev_db):
```
I think we need to catch the top level await on a task. It's not clear to me that this will work as written.

e-carlin replied:
2f7887c01

robnagler reviewed (changes requested):
Perhaps we should go over this together. I need to understand why cancelled can't be caught. We may have to solve this properly, that is, not having cancel. This might be possible.

e-carlin commented:
I squashed and updated the commit message to provide more context for what we talked about [0b60bd3](https://github.com/radiasoft/sirepo/pull/5318/commits/0b60bd3062f1447e0668daf1f87524ae5334fdf9).

robnagler reviewed (approved):
Thanks for walking me through the code.
